### PR TITLE
feat: generic progress events

### PR DESCRIPTION
**Pull request by psychedelicious**
## Summary

We don't have a way to signal progress in a generic way, so things like tiled spandrel are just an indeterminate progress bar in the UI. The UX is poor.

This PR is a first go at rectifying this. It adds a separate "generic" progress event, that accepts:
- `name: str`: used to group progress events (along with the invocation's ID)
- `step: int | None`: the current step (omit for indeterminate)
- `total_steps: int | None`: the total steps (omit for indeterminate)
- `message: str | None`: a message to display
- `image: ProgressImage | None = None`: an image to display

No frontend implementation yet as I'm still unsure of the direction of this PR.

See code comments for discussion.

## Related Issues / Discussions

Dunno if there are any.

## QA Instructions

You can test this PR by opening the network monitor in your browser's dev tools, filtering for websockets and reviewing the events while using spandrel. Start with an image > 512x512 else you only get one step.

## Merge Plan

n/a

## Checklist

- [ ] _The PR has a short but descriptive title, suitable for a changelog_
- [ ] _Tests added / updated (if applicable)_
- [ ] _Documentation added / updated (if applicable)_

### CODE REVIEW

**Comment by psychedelicious** [comment]
If we want to implement this change, once the details are stabilized I can play around with the frontend implementation. I don't anticipate it being difficult.

**Review by hipsterusername — APPROVED** [review]

**Comment by hipsterusername** [comment]
Good idea 🤘

**Comment by psychedelicious** [comment]
I've revised this PR a bit, simplifying the API and updating the frontend to use the new progress events.

Currently it's only the two spandrel nodes that signal progress with the new API. For the autoscale node, which does multiple spandrel runs, the progress bar resets for each. Its message looks like this: "Processing image (iteration 2, tile 4/25)".

While there is a message attached to each progress event, it's not yet used in the UI. I'd like to get this core implementation merged first, then iterate on the presentation.

I tried displaying the messages as a tooltip on the progress bar but it's not particularly discoverable. Maybe it should go on the invoke button tooltip? Not sure.

When that presentation is sorted, I'll update the UI's event handling to present model events as well as invocation start and complete events.

---

With this new flexible API, I'd like to explore some changes:
- Wrap `context.images.load` with progress updates, so you get some feedback as to what is happening while that long PNG encode happens.
- Currently, we have dedicated "model load" events. Maybe we should remove those and use this simpler API instead. The UI would have a single source of truth for execution status. Another option is more complicated frontend handling that updates the message with the last received event, be it a model load event or progress event. This sounds tedious.
- When downloading auxiliary models, we can use this to signal progress.

File: invokeai/app/services/shared/invocation_context.py [hunk 1]
```diff
@@ -550,13 +551,64 @@ def sd_step_callback(self, intermediate_state: PipelineIntermediateState, base_m
...
         stable_diffusion_step_callback(
-            context_data=self._data,
+            signal_progress=self.signal_progress,
             intermediate_state=intermediate_state,
             base_model=base_model,
-            events=self._services.events,
             is_canceled=self.is_canceled,
         )
 
+    def signal_progress(
```
Comment (psychedelicious) [inline invokeai/app/services/shared/invocation_context.py:560]: Maybe this should called `update_status` instead?

**Pull request closed without merging**